common/response.go

```diff
@@ -133,15 +135,20 @@ func NewHTTPResponse(ctx context.Context, req *Request, resp *network.Response,
 }
 
 func (r *Response) fetchBody() error {
+	r.bodyMx.RLock()
```

Comment:
I fear the usage of locking here can be error-prone and difficult to follow.

What about something like this?

```golang
cached := func() bool {
    r.bodyMu.Lock()
    defer r.bodyMu.RUnlock()

    return r.body != nil || r.request.frame == nil
}
if cached() {
    return nil
}
```

Reply:
I moved this fix to #124 so it can be merged quicker, and I used your suggestion there.